b9lab-content/5-my-own-chain/03-starport-13-game-wager.md

```diff
@@ -0,0 +1,271 @@
+
+Your application already includes all the necessary modules, in fact `bank` is the one you need. Let's start with letting players wager in the staking token of your application. This is the most straightforward way of doing.
+
+## New Information
+
+First, you need to add this wager value to the stored game in `proto/checkers/stored_game.proto`:
+
+```proto [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/proto/checkers/stored_game.proto#L20]
+message StoredGame {
+    ...
+    uint64 wager = 12;
+}
+```
+Plus, so as to let different player leagues emerge, you ought to let players choose the wager they want, in `proto/checkers/tx.proto`:
+
+```proto [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/proto/checkers/tx.proto#L45]
+message MsgCreateGame {
+    ...
+    uint64 wager = 4;
+}
+```
+Remember that players can still reject games, so there is no risk that players can be coerced into expensive games. To have Starport and Protobuf recompile this file, you can use:
+
+```sh
+$ starport chain serve
+```
+While you are at it, how about adding a helper function to `StoredGame`, using the Cosmos SDK's `Coin`, in `x/checkers/types/full_game.go`:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/types/full_game.go#L71-L73]
+func (storedGame *StoredGame) GetWagerCoin() (wager sdk.Coin) {
+    return sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(int64(storedGame.Wager)))
+}
+```
+Where `sdk.DefaultBondDenom` is most likely `"stake"`.
+
+## Saving The Wager
+
+Time to make sure that the new field is saved in storage and part of the creation event. First, define a new event key in `x/checkers/types/keys.go`:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/types/keys.go#L49]
+const (
+    StoredGameEventWager = "Wager"
+)
+```
+Then you need to add the field in the stored object in `x/checkers/keeper/msg_server_create_game.go`:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/keeper/msg_server_create_game.go#L29]
+storedGame := types.StoredGame{
+    ...
+    Wager: msg.Wager,
+}
+```
+And in the event:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/keeper/msg_server_create_game.go#L50]
+ctx.EventManager().EmitEvent(
+    sdk.NewEvent(sdk.EventTypeMessage,
+        ...
+        sdk.NewAttribute(types.StoredGameEventWager, strconv.FormatUint(msg.Wager, 10)),
+    )
+)
+```
+Less evidently, in order to avoid surprises down the road, you ought to also modify the creator function among the interface definition of `MsgCreateGame`, in `x/checkers/types/message_create_game.go`:
```

Comment:
```suggestion
To avoid surprises later, modify the creator function among the interface definition of `MsgCreateGame` in `x/checkers/types/message_create_game.go`:
```